File: src/0075-Sort-Colors/0075.py
class Solution:
    def _sortColors(self, nums, l, r):
        if r <= l: return
        lt = l
        gt = r + 1
        i = l + 1
        while i < gt:
            if nums[i] < nums[l]:
                lt += 1
                nums[i], nums[lt] = nums[lt], nums[i]
                i += 1
                
            elif nums[i] > nums[l]:
                gt -= 1
                nums[i], nums[gt] = nums[gt], nums[i]
                
            else:
                i += 1
        nums[l], nums[lt] = nums[lt], nums[l]
        self._sortColors(nums, l, lt - 1)
        self._sortColors(nums, gt, r)

    def sortColors(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        self._sortColors(nums, 0, len(nums) - 1)


if __name__ == "__main__":
    arr = [2,0,2,1,1,0]
    Solution().sortColors(arr)
    print(arr)

File: src/0075-Sort-Colors/0075.go
package sortcolors

func sortColors(nums []int) {
	lo, mid, hi := 0, 0, len(nums)-1
	for mid <= hi {
		if nums[mid] == 2 {
			nums[mid], nums[hi] = nums[hi], nums[mid]
			hi--
		} else if nums[mid] == 0 {
			nums[lo], nums[mid] = nums[mid], nums[lo]
			lo++
			mid++
		} else {
			mid++
		}
	}
}
